rfcs/RFC-Upload-From-Dropbox.md

```diff
@@ -0,0 +1,358 @@
+
+#### 5. Start an Instance to Download File
+
+An EC2 instance, docker container, or AWS lambda can be used to complete this task. Once a compute resource is spawned,
+it will change the upload status to uploading.
+
+#### 6. Stream File
+
+The shared link is used to stream the file from one cloud to another. The most time will be spent streaming the file
+from the DropBox to S3. DropBox offers a way to validate the integrity of a download
+by providing a hash of the data, for example [Dropbox Content Hash](https://www.dropbox.com/developers/reference/content-hash).
+As the data is streaming from the CSP, the hash of the file will be calculated as it arrives, and again as it's
+uploaded to S3. The [Content-MD5 header](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#API_PutObject_RequestSyntax)
+will be used when uploading to S3 to verify the integrity of each chunk. The AWS SDK will be useful in simplifying the
+upload to S3. For calculating the checksum while streaming, [checksumming_io](https://github.com/HumanCellAtlas/checksumming_io)
+can be used.
+
+Once the upload is complete, the download job will be removed from the queue. If the final hash does not match, then
+the upload process will be restarted. The retry logic will be an exponential backoff, with a max retry of 5 times.
+
+While streaming the file, the compute resource will periodically check if the upload has been canceled using the upload
+table. If the upload status is cancel pending, the compute resource will stop the stream, delete any data that has been uploaded,
+mark the job as complete in the queue, and set the status of the upload to cancelled.
+
+#### 7. Retry
+
+If the upload fails for any reason, the compute resource will mark the status of the upload as waiting, and return the upload
+job to the queue.
+
+#### 8. Failed Uploads
+
+If an upload fails enough times, it will be removed from the download queue. The upload status will change to failed.
+
+### Components
+
+This section describes the purpose of components that make up the upload design.
+
+#### Data Portal Browser App
+
+The DP Browser App is the frontend of the system. It is the primary method users have to interact with the Data Portal.
+It will be used to authenticate the user and interact with the DP Backend to upload files.
+
+#### Data Portal Backend
+
+The DP Backend is responsible for handling API requests. It will enqueue shared links for downloading, and update the
+upload status. This additional functionality can be built into the existing DP Backend Lambda.
+
+The backend will also be responsible for performing a simple validation of the file being uploaded. The validation is
+performed during the initial request to upload. The validation consists of verifying the extension of the file is
+".h5ad".
+
+#### Download Queue
+
+The Download Queue is an [AWS SQS](https://aws.amazon.com/sqs/) that will track pending download jobs.
```

Comment:
How long do we expect a download to take in the worst case?   We need to set the Visibility Timeout to something larger than that, or another Upload Server might pick up the task.   The default is 30 seconds which is too low in our case.   Probably a much longer time is okay, as long as we are confident that our Upload Server will be able to handle failures, and won't crash.

Reply:
We could make a pessimistic guess based on the size of the file. Initially, I'd say 15min, until we have a better idea of how long an upload takes.

Reply:
I was just reading up on Visibility Timeout.  We can update the timeout on a message at any time.  So while we are making progress on the upload, we can continuously give it more time to complete.